### PR TITLE
qa: update s3-test download code for s3-test tasks

### PR DESCRIPTION
**Pull request by alimaredia**
- Ensure the download code for all tasks running
s3-tests is consistent.
- Cleanup of download code to ensure ceph-master
runs if a specific ceph branch isn't specified
as the suite_branch

Signed-off-by: Ali Maredia <amaredia@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by yuriw** [comment]
@alimaredia did you replace all s3tests instances ?

I see this on master, in particular wondering about `mimic-x` and `nautilus-x`:

doc/changelog/v10.2.8.txt:    s3tests: add force-branch with higher precdence than 'branch'
doc/releases/mimic.rst:* rgw,tests: qa/tasks/swift: add support for the "force-branch" configurable (`pr#21027 <https://github.com/ceph/ceph/pull/21027>`_, Radoslaw Zarzynski)
qa/suites/rgw/crypt/2-kms/barbican.yaml:      force-branch: master
qa/suites/rgw/crypt/2-kms/barbican.yaml:      force-branch: master
qa/suites/rgw/crypt/4-tests/s3tests.yaml:      force-branch: ceph-master
qa/suites/rgw/multifs/tasks/rgw_s3tests.yaml:      force-branch: ceph-master
qa/suites/rgw/tempest/tasks/rgw_tempest.yaml:      force-branch: master
qa/suites/rgw/tempest/tasks/rgw_tempest.yaml:      force-branch: master
qa/suites/rgw/thrash/workload/rgw_s3tests.yaml:      force-branch: ceph-master
qa/suites/rgw/verify/tasks/s3tests-java.yaml:        force-branch: master
qa/suites/rgw/verify/tasks/s3tests.yaml:      force-branch: ceph-master
qa/suites/rgw/website/tasks/s3tests-website.yaml:      force-branch: ceph-master
qa/suites/teuthology/rgw/tasks/s3tests-civetweb.yaml:      force-branch: master
qa/suites/teuthology/rgw/tasks/s3tests-fastcgi.yaml:      force-branch: master
qa/suites/teuthology/rgw/tasks/s3tests-fcgi.yaml:      force-branch: master
qa/suites/upgrade/mimic-x/parallel/5-final-workload/rgw_swift.yaml:        force-branch: ceph-master
qa/suites/upgrade/nautilus-x/parallel/5-final-workload/rgw_swift.yaml:        force-branch: ceph-master
qa/tasks/barbican.py:        branch = cconf.get('force-branch', 'master')
qa/tasks/barbican.py:            force-branch: master
qa/tasks/barbican.py:            force-branch: master
qa/tasks/barbican.py:            force-branch: master
qa/tasks/cbt.py:        branch = self.config.get('force-branch', branch)
qa/tasks/keystone.py:                '-b', cconf.get('force-branch', 'master'),
qa/tasks/keystone.py:            force-branch: master
qa/tasks/ragweed.py:        branch = cconf.get('force-branch', None)
qa/tasks/s3readwrite.py:        branch = cconf.get('force-branch', None)
qa/tasks/s3roundtrip.py:        branch = cconf.get('force-branch', None)
qa/tasks/s3tests.py:        branch = cconf.get('force-branch', None)
qa/tasks/s3tests_java.py:                force-branch: wip
qa/tasks/s3tests_java.py:            if 'force-branch' in self.config[client] and self.config[client]['force-branch'] is not None:
qa/tasks/s3tests_java.py:                branch = self.config[client]['force-branch']
qa/tasks/swift.py:                '-b', cconf.get('force-branch', 'ceph-master'),
qa/tasks/tempest.py:                '-b', cconf.get('force-branch', 'master'),
qa/tasks/tempest.py:            force-branch: master

**Comment by alimaredia** [comment]
@yuriw afaik yes. Those mimic-x and nautilus-x grep lines you see here:
`qa/suites/upgrade/mimic-x/parallel/5-final-workload/rgw_swift.yaml: force-branch: ceph-master
qa/suites/upgrade/nautilus-x/parallel/5-final-workload/rgw_swift.yaml: force-branch: ceph-master`
are for swift tests not s3-tests. Many non-s3tests tasks use `force-branch: ceph-master.

**Comment by smithfarm** [comment]
@alimaredia Could you add the following line to the commit message:

```
Fixes: https://tracker.ceph.com/issues/43077
```

Also, it seems like https://github.com/ceph/s3-tests has `ceph-*` branches for all the named branches, not just master. So I'm guessing it would make sense to backport this to nautilus and mimic, at least. Do you agree?

**Comment by yuriw** [comment]
test this please

**Comment by alimaredia** [comment]
@cbodley I changed it back to say `force-branch`.

@smithfarm Yes, that is the plan. Once this PR gets merged, I am planning on creating PRs to the nautilus, mimic, and luminous branches with `ceph-master` changed to `ceph-luminous`. If you'd do it differently, let me know.

**Comment by alimaredia** [comment]
@cbodley @yuriw: here is a teuthology run with this branch as the suite branch: http://pulpito.ceph.com/amaredia-2019-12-05_22:47:18-rgw:verify-master-distro-basic-smithi/

File: qa/tasks/s3readwrite.py [hunk 1]
```diff
@@ -18,29 +18,32 @@
...
     log.info('Downloading s3-tests...')
     testdir = teuthology.get_testdir(ctx)
-    for (client, cconf) in config.items():
-        branch = cconf.get('force-branch', None)
-        if not branch:
-            branch = cconf.get('branch', 'master')
-        sha1 = cconf.get('sha1')
+    for (client, client_config) in config.items():
+        s3tests-branch = client_config.get('force-branch', None)
+        if not s3tests-branch:
```
Comment (cbodley) [inline qa/tasks/s3readwrite.py]:
```
  File "/home/teuthworker/src/github.com_alimaredia_ceph_wip-s3-test-branch-refactor/qa/tasks/s3tests.py", line 35
    s3tests-branch = client_config.get('force-branch', None)
SyntaxError: can't assign to operator
```
variable names can't have dashes in them

File: qa/tasks/s3roundtrip.py [hunk 2]
```diff
@@ -18,35 +18,48 @@
...
-            branch = cconf.get('branch', 'master')
+    for (client, client_config) in config.items():
+        s3tests-branch = client_config.get('force-branch', None)
+        if not s3tests-branch:
+            raise ValueError(
+                "Could not determine what branch to use for s3-tests. Please add 'force-branch: {s3-tests branch name}' to the .yaml config for this s3roundtrip task.")
+
+        log.info("Using branch '%s' for s3tests", s3tests-branch)
+        sha1 = client_config.get('sha1')
+        git_remote = client_config.get('git_remote', None) or teuth_config.ceph_git_base_url
```
Comment (tchaikov) [inline qa/tasks/s3roundtrip.py]:
```suggestion
        git_remote = client_config.get('git_remote', teuth_config.ceph_git_base_url)
```

**Comment by tchaikov** [comment]
@alimaredia ping?

**Comment by alimaredia** [comment]
@tchaikov took your latest recommendation and the issues Casey pointed out. 

Started a new run here: http://pulpito.front.sepia.ceph.com/amaredia-2019-12-10_17:41:02-rgw:verify-master-distro-basic-smithi/

**Comment by alimaredia** [comment]
@cbodley @tchaikov: the above run (http://pulpito.ceph.com/amaredia-2019-12-10_17:41:02-rgw:verify-master-distro-basic-smithi/) worked. AFAIK this code should be good to merge.

**Comment by yehudasa** [comment]
@alimaredia @cbodley @tchaikov I may be late to the party here, but the idea that we need to update the suites force-branch for every version is not sustainable. Already missed the `s3readwrite` and `s3roundtrip` tasks. Use of `force-branch` was designed to to override implicit branch resolving, not to completely replace it.
IMO we should do this:
 - if `force-branch` is defined, use it
 - else if exists branch `ceph-$ceph_branch` use that
 - else if exists branch `$ceph_branch` use that
 - else:
 - read current ceph_release off `$ceph_repo/src/ceph_release`
 - if exists `ceph-$ceph_release` branch use that
 - else use `ceph-master`

**Comment by alimaredia** [comment]
@yehudasa I added a PR for suites using the `s3readwrite` and `s3roundtrip` tasks in https://github.com/ceph/ceph/pull/32225. 

I wanted to rewrite the download code to make it as simple as possible. If there is an issue with the download code the teuthology logs should show why. If someone trying to debug the issue looked at the code they shouldn't spend any more than a minute understanding where the issue took place.

When I began writing this PR, I thought I would just have to change a line or a variable. From there I noticed that the download code was not the same for `s3roundtrip`, `s3readwrite`, and `s3tests`. Since the code was doing the same thing I thought it should be the same for all 3. The way I saw it, branch prediction looked similar to how you outlined it should look. 

After I discussed it with Casey, we both agreed it would be simpler to do away with branch resolving entirely. The way I originally had the PR, `force-branch` was really `s3tests-branch`. Casey wanted to use `force-branch` to match other tasks.

If you've got a good argument for branch resolving, I'd be open to hearing it. The most important thing for me is that the code is easy to understand and consistent across tasks and releases.

**Pull request merged**